moveit_kinematics/kdl_kinematics_plugin/src/kdl_kinematics_plugin.cpp

```diff
@@ -367,58 +433,102 @@ bool KDLKinematicsPlugin::searchPositionIK(const geometry_msgs::Pose& ik_pose, c
   return false;
 }
 
-int KDLKinematicsPlugin::CartToJnt(KDL::ChainIkSolverVel& ik_solver, const KDL::JntArray& q_init,
-                                   const KDL::Frame& p_in, KDL::JntArray& q_out, const unsigned int max_iter) const
+int KDLKinematicsPlugin::CartToJnt(KDL::ChainIkSolverVelMimicSVD& ik_solver, const KDL::JntArray& q_init,
+                                   const KDL::Frame& p_in, KDL::JntArray& q_out, const unsigned int max_iter,
+                                   const Eigen::VectorXd& joint_weights, const Twist& cartesian_weights) const
 {
+  double last_delta_twist_norm = DBL_MAX;
+  double step_size = 1.0;
   KDL::Frame f;
   KDL::Twist delta_twist;
-  KDL::JntArray delta_q(q_out.rows());
+  KDL::JntArray delta_q(q_out.rows()), q_backup(q_out.rows());
+  Eigen::ArrayXd extra_joint_weights;
+  extra_joint_weights.setOnes(joint_weights.rows());
 
   q_out = q_init;
   ROS_DEBUG_STREAM_NAMED("kdl", "Input: " << q_init);
 
   unsigned int i;
+  bool success = false;
   for (i = 0; i < max_iter; ++i)
   {
     fk_solver_->JntToCart(q_out, f);
     delta_twist = diff(f, p_in);
     ROS_DEBUG_STREAM_NAMED("kdl", "[" << std::setw(3) << i << "] delta_twist: " << delta_twist);
 
-    if (position_ik_)
+    // check norms of position and orientation errors
+    const double position_error = delta_twist.vel.Norm();
+    const double orientation_error = ik_solver.isPositionOnly() ? 0 : delta_twist.rot.Norm();
+    const double delta_twist_norm = std::max(position_error, orientation_error);
+    if (delta_twist_norm <= epsilon_)
     {
-      if (fabs(delta_twist(0)) < epsilon_ && fabs(delta_twist(1)) < epsilon_ && fabs(delta_twist(2)) < epsilon_)
-        break;
+      success = true;
+      break;
+    }
+
+    if (delta_twist_norm >= last_delta_twist_norm)
+    {
+      // if the error increased, we are close to a singularity -> reduce step size
+      double old_step_size = step_size;
+      step_size *= std::min(0.5, last_delta_twist_norm / delta_twist_norm);  // reduce scale;
```

Comment:
I figured using a lower min value (i.e. 0.2) reduces the time being stuck at singularities.
This can be related to the setup, but is there another reason why this shouldn't be lower?

Reply:
This is a rather uncritical value. Choosing it too high, simply requires several iterations to actually decrease the error. Choosing it too small, might unnecessarily slow down convergence.
Everything staying in the same order of magnitude is fine. I will try myself on our robot zoo and then merge.